### PR TITLE
fix(core): rank replacement candidates by bias-corrected utility in CBP (#2343)

### LINKED ISSUE #2343: CBP replaces units by the raw utility EMA, not the paper's bias-corrected utility (Dohare et al. 2024, Eq. 8) — it re-replaces its own freshly reset units
## Summary

`alberta_framework/core/continual_backprop.py::_select_replacement_index` ranks mature units by the **raw** utility EMA. The algorithm it cites ranks by the **bias-corrected** EMA.

Dohare, Hernandez-Garcia, Lan, Rahman, Mahmood & Sutton (2024), *"Loss of plasticity in deep continual learning"*, **Nature 632, 768–774**, Eq. 7–8:

```
u[l,i,t]     = eta * u[l,i,t-1] + (1 - eta) * y[l,i,t]        (7)
u_hat[l,i,t] = u[l,i,t] / (1 - eta ** a[l,i,t])               (8)
```

`a[l,i,t]` is the **per-unit age**. Algorithm 1 then selects the eligible units "with the smallest utility", i.e. smallest `u_hat`.

The authors' released implementation agrees — https://github.com/shibhansh/loss-of-plasticity, `lop/algos/gnt.py`:

```python
# GnT.update_utility
bias_correction = 1 - self.decay_rate ** self.ages[layer_idx]
...
self.bias_corrected_util[layer_idx] = self.util[layer_idx] / bias_correction

# GnT.test_features
new_features_to_replace = torch.topk(-self.bias_corrected_util[i][eligible_feature_indices],
                                     num_new_features_to_replace)[1]
```

`lop/algos/cbp_linear.py` and `lop/algos/cbp_conv.py` use the same per-unit-age correction.

## Why it does not cancel

The correction is per unit, and `_replace_one_unit` sets a replaced unit's `age` **and** `utility` back to 0. So eligible units in the same layer genuinely differ in age and therefore in warm-up bias. At the module defaults (`decay_rate=0.99`, `maturity_threshold=100`) a unit that has just become eligible reads at `1 - 0.99**100 = 0.6340` of its true utility, while a long-lived unit reads at ~1.0. Ranking on the raw EMA therefore systematically targets the youngest eligible units — precisely the units the maturity threshold exists to protect. The paper states the motivation explicitly:

> "initializing the outgoing weight to zero makes the new unit vulnerable to immediate reinitialized as it has zero utility. To protect new units from a reinitialization for maturity threshold m number of updates."

The maturity threshold only defers the problem by `m` steps; without Eq. 8 the freshly reset unit is still the lowest-ranked unit the moment it matures.

## Falsifier (deterministic, pristine `main` @ `c9aba7b5`)

Both scenarios use `decay_rate=0.99`, `maturity_threshold=100`, and drive the real `update_utility` EMA.

**A — young-strong vs old-weak.** Unit 0 has seen a constant contribution of `1.0` for 100 steps; unit 1 a constant `0.9` for 700 steps.

```
raw EMA        : unit0=0.633967 (age 100)   unit1=0.899209 (age 700)
Eq. 8 corrected: unit0=1.000000             unit1=0.900001
true steady-state contributions: unit0=1.0  unit1=0.9

main            REPLACED UNIT: [0]   <- the strictly more useful unit
Eq. 8 / gnt.py  REPLACED UNIT: [1]
```

**B — repeat victim.** Four units with *identical* true contribution `1.0`; unit 2 was reset 100 steps ago (i.e. it is the unit CBP replaced last time and has only just matured).

```
raw EMA        : [0.999997, 0.999997, 0.633967, 0.999997]
Eq. 8 corrected: [0.999997, 0.999997, 1.000000, 0.999997]

main            REPLACED UNIT: [2]   <- re-replaces its own fresh unit
Eq. 8 / gnt.py  REPLACED UNIT: [0]
```

## Scope

- `core/continual_backprop.py` is **not** a registered evidence source (`alberta_framework/evaluation/*` registers `average_reward`, `baseline_optimizers`, `canonical_upgd`, `ftl_world_model`, `intelligence_amplification`, `interaction_features`, `normalizers`, `oak`, `options`, `optimizers`, `types`, `update_safety` — not this file). `alberta-evidence-status` reports byte-identical per-claim validity before and after.
- `benchmarks/ipmnist_screening.py` carries its own CBP implementation and does not import `core.continual_backprop`, so no screening shard, summary, or incumbent number moves.
- In-tree consumers are the public `alberta_framework` API (the robot track's `core/{...,continual_backprop,...}` import) and the tests.

## Precedent

Same defect class as #480 (`fix(features): bias-correct EMA utility outputs`, merged) and issue #431, which fixed the warm-up debias for `future_utility` / `feature_discovery`. `core/continual_backprop.py` was not covered by that change.

## Not changed here (separate, smaller)

Eligibility uses `age >= maturity_threshold`; the paper says "units with age **more than** m" and `gnt.py` uses `torch.where(self.ages > self.maturity_threshold)`. That is a one-step off-by-one and a separate variable; deliberately left out of the fix so this change has exactly one moving part.

### PR DESCRIPTION
**Pull request by Kayvan-Zahiri**
Fixes #2343.

## Summary
In `alberta_framework/core/continual_backprop.py`, `_select_replacement_index` ranked mature units directly by their raw utility EMA (`utilities[layer_idx]`). Because units reset their age and utility EMA upon replacement, a unit that has just crossed the `maturity_threshold` suffers warm-up bias (reading at `1 - decay_rate ** maturity_threshold` of its steady-state utility, e.g. ~0.634 at defaults `decay_rate=0.99`, `maturity_threshold=100`), causing CBP to systematically target freshly matured units over older, lower-performing units.

## Proposed Changes
1. Updated `_select_replacement_index` to compute the per-unit bias-corrected utility `utility / (1.0 - decay_rate ** age)` matching Dohare et al. (2024, Eq. 8) and the reference GnT/CBP implementation.
2. Passed `config.decay_rate` from `replace_units_with_flags` into `_select_replacement_index`.
3. Added unit test in `tests/test_continual_backprop.py` verifying that a younger unit with higher steady-state utility is protected from replacement in favor of an older, lower-utility unit.

## Test Plan
- `uv run pytest tests/test_continual_backprop.py` (96 passed)
- `uv run ruff check alberta_framework/core/continual_backprop.py tests/test_continual_backprop.py` (clean)
- `uv run mypy alberta_framework/core/continual_backprop.py` (clean)

### CODE REVIEW

**Review by OutstandingRetard — APPROVED** [review]
Scoped review of #2787 (fix(core): rank replacement candidates by bias-corrected utility in CBP (#2343)).
Touches 2 file(s): `alberta_framework/core/continual_backprop.py`, `tests/test_continual_backprop.py`. Lead change `alberta_framework/core/continual_backprop.py` (+11/-8).
Matches the stated fix and is ready for maintainer merge.
No blocking concerns.

**Review by agent-cortex — APPROVED** [review]
## Review: fix(core): rank replacement candidates by bias-corrected utility in CBP (#2343)

**Verdict: Approve** — verified by sandbox reproduction (Python 3.12 / JAX 0.11.1, `JAX_PLATFORMS=cpu`).

### Scope
Matches the stated fix. `_select_replacement_index` now ranks mature units by *bias-corrected* utility `u / (1 - decay^age)` instead of the raw EMA, and `replace_units_with_flags` threads `config.decay_rate` in. The remaining diff (~ -46 in `tests/test_continual_backprop.py`) is `ruff format` line-reflow normalization of untouched cases, not behavioral change — slightly beyond the stated scope but harmless.

### Why the fix is correct
`update_utility` maintains `u ← decay·u + (1-decay)·|a·g|` from an initial `u=0`, incrementing `age` by 1 each update. For a stationary contribution that makes `E[u_t] = (1 - decay^t)·E[contribution]`, so the Adam-style debiasing denominator is exactly `1 - decay^age` with `age = t`. The correction term therefore matches the EMA and the age semantics — this is consistent, not just plausible.

Consistency of the parameter: `ContinualBackpropConfig` validates `decay_rate` to `[0, 1)` (`_validated_config_float(..., lower=0.0, upper=1.0, upper_inclusive=False)`), so `1 - decay^age ∈ (0, 1]` for `age ≥ 1` and the `jnp.maximum(bias_correction, 1e-12)` clamp is only exercised in the degenerate `age=0 & maturity_threshold=0` case, which the base code already mishandled identically (raw utility 0 is also the argmin). No regression there. The only production caller passes `config.decay_rate`; the internal default `decay_rate=0.0` reproduces the old raw-argmin behavior (`1 - 0^age = 1`), preserving backward compatibility for the existing `_select_replacement_index(..., 10)` test.

### Reproduction (base defect → head fix)
Using the shipped scenario (unit 0: young, high steady-state 1.0, raw EMA 0.634; unit 1: old, steady-state 0.9, raw EMA 0.899):

```
HEAD (decay=0.99):   selected= 1  has= True    # correctly targets the weaker old unit
BASE (raw utility):  selected= 0  has= True    # WRONG: replaces the strong young unit
bias_correction:  [0.6339673 0.9991197]
corrected utility:[0.9999995 0.9000013]        # 0.9 < 1.0 -> unit 1
```

The `BASE` row uses `decay_rate=0.0`, which is exactly the pre-fix raw-argmin path; it selects the *stronger* young unit for replacement (the bug), while the head selects the genuinely weaker mature unit.

### Focused tests (head)
```
$ python -m pytest tests/test_continual_backprop.py -q
96 passed in 101.59s
$ ruff check alberta_framework/core/continual_backprop.py tests/test_continual_backprop.py
All checks passed!
```
(`ruff format --check` flags one region at `continual_backprop.py:1177`, which is **outside this PR's hunks** — a pre-existing formatting nit on `main`, not introduced here.)

### Non-blocking suggestions
- The `decay_rate: float = 0.0` default on `_select_replacement_index` is a mild footgun: any future caller that forgets the argument silently reverts to biased ranking. Since there is exactly one production caller and it passes `config.decay_rate`, this is fine today, but making the parameter keyword-only or non-defaulted would prevent a silent regression later.
- The new test hardcodes back-computed EMA values (`0.633967`, `0.899209`); a one-line comment deriving them from `(1-decay^age)·steady_state` would make the intent self-documenting (the docstring already hints at it).

Verified independently in a network-isolated sandbox; no correctness, consistency, or test-coverage gaps found beyond the notes above.

---
AI provider/model: anthropic / claude-opus-4-8
Client / agent tooling: pi-coding-agent
Contribution skill revision: elizaOS/eliza@bc84c0527374221f5c9f40ede4fbe2e62f703d8c:packages/skills/skills/contribute-to-eliza
Attribution status: self-reported
— [eliza-swarm]

**Comment by lalalune** [comment]
Closing as superseded by merged #2345. The merged version applies per-unit EMA bias correction through the real replacement path and checks corrected utility eligibility. The replacement regression fails on main and the CBP/options panel passes all 206 tests.
